sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala

```diff
@@ -105,13 +105,15 @@ abstract class StringRegexExpression extends BinaryExpression
     Examples:
       > SELECT _FUNC_('Spark', '_park');
       true
+      > SELECT '\\abc' AS S, S _FUNC_ R'\\abc', S _FUNC_ '\\\\abc';
+      \abc	true	true
       > SET spark.sql.parser.escapedStringLiterals=true;
       spark.sql.parser.escapedStringLiterals	true
       > SELECT '%SystemDrive%\Users\John' _FUNC_ '\%SystemDrive\%\\Users%';
       true
       > SET spark.sql.parser.escapedStringLiterals=false;
       spark.sql.parser.escapedStringLiterals	false
-      > SELECT '%SystemDrive%\\Users\\John' _FUNC_ '\%SystemDrive\%\\\\Users%';
```

Comment:
This is the example you want to fix, right?

Reply:
Do we need to change this part?

https://github.com/apache/spark/blob/fdedec1de54fa7619e1d1b59eb6f15a49e727428/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala#L95-L96

Reply:
it looks to me that we follow the java string semantic: if you write `"\\abc"`, you actually get `\abc`. Is it also true in simple `SELECT` not `LIKE`?

Reply:
Do we have a doc to explain what the string value we get eventually by writing string literals in SQL?

Reply:
I'm seeing some weird behaviors:
```
scala> sql(""" select "a\b" """).show
+---+
|a\b|
+---+
|a\b|
+---+

scala> sql(""" select "a\\b" """).show
+---+
|a\b|
+---+
|a\b|
+---+
```

Reply:
> https://github.com/apache/spark/blob/fdedec1de54fa7619e1d1b59eb6f15a49e727428/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala#L95-L96

I guess the "\abc" means the fact value but the literal. Of course, it make users confused. I agree to update the doc here to give more explanation.

Reply:
> Do we have a doc to explain what the string value we get eventually by writing string literals in SQL?

There is the doc about string literals: https://spark.apache.org/docs/latest/sql-ref-literals.html#string-literal. It explain something implemented in https://github.com/apache/spark/blob/5f643ee1444d41ca78ef397d9675efac550a6967/sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkParserUtils.scala#L34-L54

Reply:
> This is the example you want to fix, right?
> SELECT '%SystemDrive%\\Users\\John' _FUNC_ '\%SystemDrive\%\\\\Users%';

@dongjoon-hyun I would change this to don't mix 2 different things: how the SQL parser handles regular string literals and how `LIKE` handles patterns. Not obvious why those 2 escapes `/` and `\` lead to 2 different patterns:
```sql
spark-sql (default)> SELECT R'%SystemDrive%\Users\John' LIKE '\%SystemDrive\%\\\\Users%' ESCAPE R'\';
true
spark-sql (default)> SELECT R'%SystemDrive%/Users/John' LIKE '/%SystemDrive/%//Users%' ESCAPE R'/';
true
```

Reply:
LIKE and Java both handle `\` and only LIKE handles `/`.

Reply:
> Use \ to escape special characters

So for `\a` that doesn't escape anything, `\` is treated as a normal char.

I think the problem here is the LIKE doc describes the behavior with the actual string value of the pattern. Can we reference the string literal doc here? And give examples about how to match `\abc`: `r'\\abc'` or `'\\\\abc'`.

Reply:
I have updated the description of string literals in https://github.com/apache/spark/pull/43152, so, we can refer to the doc from here and other expression descriptions. Please, take a look at the PR.